File: ListNode/ListNode.cpp
#include <unordered_map>
using namespace std;
struct ListNode {
    int key;
    int value;
    ListNode *next, *prev;

    ListNode() : key(0), value(0), next(nullptr), prev(nullptr) {}
    ListNode(int key, int value) : key(key), value(value), next(nullptr), prev(nullptr) {}
    ListNode(int key, int value, ListNode* next, ListNode* prev) : key(key), value(value), next(next), prev(prev) {}
    ~ListNode() {delete next, prev;}
};

class LRUCache {
public:

    LRUCache(int capacity) {
        size = capacity;
        head = new ListNode();
        tail = new ListNode();
        head->next = tail, tail->prev = head;
    }
    ~LRUCache() {
        while (head != tail) {
            head = head->next;
            delete head->prev;
        }
        delete head;
    }
    
    int get(int key) {
        if (memo.count(key)){
            _remove(memo[key]);
            _insert(memo[key]);
            return memo[key]->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        ListNode *node;
        if (memo.count(key) || size == 0) {
            node = memo.count(key) ? memo[key] : tail->prev;
            memo.erase(node->key);
            _remove(node);
            node->key = key;
            node->value = value;
        } else {
            node = new ListNode(key, value);
            size--;
        }
        memo[key] = node;
        _insert(node);
    }
    
    void _remove(ListNode *node) {
        ListNode *prev = node->prev, *next = node->next;
        prev->next = next;
        next->prev = prev;
    }

    void _insert(ListNode *node) {
        node->prev = head;
        node->next = head->next;

        head->next->prev = node;
        head->next = node;
    }
private:
    int size=0;
    ListNode *head, *tail;
    unordered_map<int, ListNode*> memo;
};

File: ListNode/ListNode.py
from collections import defaultdict

class ListNode:
    def __init__(self, key=0, value=0, next=None, prev=None) -> None:
        self.key = key
        self.value = value
        self.next = next
        self.prev = prev
    

class LRUCache:
    def __init__(self, capacity=float('inf')) -> None:
        assert capacity > 0, f"Should has capacity > 0, but get {capacity}"

        self.head = ListNode()
        self.tail = ListNode()
        self.head.next = self.tail
        self.tail.prev = self.head

        self.capacity = capacity
        self.memo = defaultdict(ListNode)
    
    def get(self, key : int):
        if key in self.memo:
            node = self.memo[key]
            self._remove(node)
            self._insert(node)
            return node.value
        else:
            return -1
    
    def put(self, key, value):
        
        if key in self.memo:
            node = self.memo[key]
            self._remove(node)
            self.memo[key] = node.value = value
        else:
            if self.capacity == 0:
                node = self.tail.prev
                del self.memo[node.key]
                node.key, node.value = key, value
                self._remove(node)
            else:
                node = ListNode(key, value)
                self.capacity -= 1
            self.memo[key] = node
        self._insert(node)
    
    def _remove(self, node: ListNode) -> ListNode: 
        prev = node.prev
        next = node.next

        prev.next = next
        next.prev = prev
        return node
    
    def _insert(self, node: ListNode):
        node.prev = self.head
        node.next = self.head.next

        self.head.next = node
        node.next.prev = node
        return node